draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1439,158 +1488,237 @@ Steps:
-among multiple invocations of hash\_to\_base, i.e., by factoring out the common
-computation.
-
-## Implementation {#hashtobase-impl}
-
-The following procedure implements hash\_to\_base.
+recommendations for choosing DST, the domain separation tag.
+This is the REQUIRED method for applying domain separation.
 
 ~~~
-hash_to_base(msg, ctr)
+hash_to_field(msg, count)
 
 Parameters:
 - DST, a domain separation tag (see discussion above).
-- H, a cryptographic hash function.
 - F, a finite field of characteristic p and order q = p^m.
 - p, the characteristic of F (see immediately above).
 - m, the extension degree of F, m >= 1 (see immediately above).
 - L = ceil((ceil(log2(p)) + k) / 8), where k is the security
   parameter of the cryptosystem (e.g., k = 128).
-- HKDF-Extract and HKDF-Expand are as defined in RFC5869,
-  instantiated with the hash function H.
+- expand_message, a function that expands an octet-string and
+  domain separation tag into a pseudorandom octet-string
+  (see discussion above).
 
 Inputs:
-- msg is the message to hash.
-- ctr is 0, 1, or 2.
-  This is used to efficiently create independent
-  instances of hash_to_base (see discussion above).
+- msg is an octet-string containing the message to hash.
+- count is the number of elements of F to output.
+
+Outputs:
+- (u_1, ..., u_(count)), a list of field elements.
+
+Notation:
+- For an octet-string str, let str[a : b] represent the
+  (b - a) octets starting at the a'th octet of str.
+
+Steps:
+1. pro_length = count * m * L
+2. pseudo_random_octets = expand_message(msg, DST, prb_length)
+3. for i in (0, ..., count - 1):
+4.   for j in (0, ..., m - 1):
+5.     elm_offset = L * (j + i * m)
+6.     tv = pseudo_random_octets[elm_offset : (elm_offset + L)]
+7.     e_j = OS2IP(tv) mod p
+8.   u_i = (e_0, ..., e_(m - 1))
+9. return (u_0, ..., u_(count - 1))
+~~~
+
+## expand\_message {#hashtofield-expand}
+
+expand\_message is a function that generates a pseudorandom octet-string.
+It takes three arguments:
+
+- msg, an octet-string containing the message to hash,
+- DST, an octet-string that acts as a domain separation tag, and
+- len\_in\_octets, the number of octets to be generated.
+
+This document defines two variants of expand\_message:
+
+- expand\_message\_md ({{hashtofield-expand-md}}) is appropriate for use
+with a wide range of hash functions, including SHA-2 {{FIPS180-4}}, SHA-3
+{{FIPS202}}, BLAKE2 {{?RFC7693}}, and others.
+
+- expand\_message\_xof ({{hashtofield-expand-xof}}) is appropriate for use
+with extensible-output functions (XOFs) including functions in the SHAKE
+{{FIPS202}} or BLAKE2X {{BLAKE2X}} families.
+
+These two variants should suffice for the vast majority of use cases, but other
+variants are possible; {{hashtofield-expand-other}} discusses requirements.
+
+### expand\_message\_md {#hashtofield-expand-md}
+
+The expand\_message\_md function produces a pseudorandom octet string using
+a cryptographic hash function H that outputs b bits.
```

Comment:
Should we elaborate on why this is indifferentiable from a RO? Perhaps we can cite [1] and note that we encode the input `msg` in a prefix-free manner (by prepending the length and DST)? (And whatever else is needed?)

[1] https://cs.nyu.edu/~dodis/ps/merkle.pdf

Reply:
OK, I've added discussion on this in Security Considerations.

(Does that seem like the right spot for it?)

Reply:
Yep! I'll take a look.